### PR TITLE
Remove the use of cron to monitor jobs

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The job monitoring is now done on the fly when ever a non-terminated job is loaded. The following changes have been made:

* `load_job` and `load_all` now run the `monitor` for non-terminated jobs,
* The `cron` related ruby code has been removed,
* `libexec/check-cron` and the original `libexec/run-monitor` have been removed,
* The `libexec/run-monitor` will now remove legacy `crontab` for the user,
* `transition-inactive` is only ran when indexing `jobs`, and
* A `--wait` flag has been added to `info-job` which hangs until the job terminates.

---

The `libexec/run-monitor` shim has been repurposed to uninstall the existing crontab. This works because `cron` has been configured with the absolute path to the shim. This makes removing the crontab entry relatively straight forward.

This is not done within the `post-inst` scripts because `root` does not know which user's have a `crontab` entry. Any `post-inst` script would have to loop through `/etc/passwd`, making the process needlessly complicated.

---

The `transition-inactive` utility use be ran by the monitoring process and needs to be replaced. This is used to detect if a job's submission failed and never writes its metadata.

A few changes have been made to this:

* The `active_path` has been renamed `initial_metadata_path`,
* The file is created just before running `submit.sh`,
* The `transition_inactive` is solely responsible for removing the initial metadata path, and
* The `list-jobs` command will index the `initial_metadata_path` and run the `transition_inactive`.

Previously multiple locations would try and delete the `initial_metadata_path` files, making it hard to follow the life-cycle of the file. Instead the `list-jobs`/`transition_inactive` utility is solely responsible for removing the file. One of the following will occur:

* If `metadata_path` was created successfully, delete `initial_metadata_path` (to be expected),
* If the `submission_period` has not elapsed, do nothing
* If the `submission_period` has elapsed, log an error, transition the job to a `FAILED` state, and delete the file, or
* If `initial_metadata_path` is invalid then the job is un-recoverable! This should never happen in practice. Log the error and delete the file.

---

~A new `--wait` flag has been added to the `info-job` command. This will cause the command to hang until the job terminates. There is no timeout as part of this command, this will need to be implemented by the calling process if required.~

~Instead `info--job` will start running the `monitor` script within a loop. There is a `wait_cooldown` period within the loop to prevent spamming calls to `scontrol`.~

Not applicable: The `--wait` flag has been dropped.

### CODE REVIEW

**Comment by WilliamMcCumstie** [comment]
Note to future self, the `--wait` flag can be removed.

It is being reworked in a different PR

**Comment by benarmston** [comment]
@WilliamMcCumstie a quick glance through this and the changes to monitor the jobs automatically look good.  However, on further consideration of this I think we need to have the `run-monitor` command back.  Consider the following:

* A user submits a job.
* Time passes, the job starts, completes and is moved to the accounting record.
* Lots more time passes and the job is removed from the from accounting record.
* The user runs `info-job` and is informed that their job is still `pending`.

This doesn't look good.

I suspect that this is a very unlikely thing to happen and if sufficient time passes without the user running `info-job`/`list-jobs`, they probably don't care about that job any longer.  However, as we already have code to deal with this I think it is something that we can have a "solution" for.  That solution is as follows:

1. We provide a mechanism to update the job state.  I.e., `/opt/flight/bin/flight job run-monitor`.
2. The cluster admin (or ops team / integrator) writes a standard cronjob which runs the above for whatever users they have identified, with a frequency they have identified.

I suspect that there won't be any admins that care about this use case, but its good that we have a solution if they do.

**Comment by WilliamMcCumstie** [comment]
@benarmston FYI, the above assessment about jobs getting stuck in a `PENDING` state isn't entirely correct. The actual behaviour depends on how `libexec/<scheduler>/monitor.sh` handles "missing jobs".

In the case of `slurm`, there is already some logic which will transition the job to the `UNKNOWN` state if `sacct` has lost track of the job:

```
  # Transition the job to "UNKNOWN" is sacct has no record of it
  if [[ "$exit_status" -eq 0 ]] && [ -z "$acct" ]; then
    state="UNKNOWN"
    start_time=''
    end_time=''
    reason=''
```

So in the above situation, the slurm jobs will transitioned from `PENDING` to `UNKNOWN`. This behaviour should be replicated for all other schedulers as well, however failure to do so will result in the "Stuck PENDING" edge-case.

---

In general, getting jobs into the `UNKNOWN` state should be avoided, so I have added the `run-monitor` command back regardless.

**Comment by WilliamMcCumstie** [comment]
NOTE: This branch has been rebased back onto the current state of `master` to resolve conflicts.

A new `release/2.3.1` branch contains the version which was rebased onto the `2.3.0` tag.

**Comment by benarmston** [comment]
Rebased on latest `master`.

**Review by benarmston — COMMENTED** [review]
I have a question about the `run_monitor` command.  Other than that, this looks good.

File: lib/flight_job/commands/run_monitor.rb [hunk 1]
```diff
@@ -29,13 +29,8 @@ module FlightJob
...
       def run
-        Job.load_active.each do |job|
-          if job.valid?(:monitor)
-            job.monitor
-          else
-            job.transition_inactive
-          end
-        end
+        Job.transition_inactive
+        Job.load_all
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/flight_job/commands/run_monitor.rb:33]: The `run_monitor` command has two purposes:

1. Pick up jobs where `submit.sh` failed (aka crashed an burned), and
2. Run the monitor script for all non-terminal jobs.

---

## Use case 1: Transition inactive

Use case one is fairly easy, glob for all the `active` files and check the scheduler if it knows anything about it. It then runs some logic to determine if the job was submitted to the scheduler or if it failed.

Regardless, the initial_metadata_path _should_ be deleted after `Job#transition_inactive` is called.

*NOTE:* `active_path` has been renamed as it is rather obscure in meaning. `initial_metadata_path` is far more descriptive. The are still some methods which use the `active`/`inactive` terminology.

---

## Use case 2: Update Non-Terminal Jobs

The secondary use case is to update the metadata state for non-terminal jobs. This is a bit trickier as the state is stored within the `metadata`. Currently the only way to check this `state` is to load all the `jobs`. This will trigger the `monitor.sh` to be ran (when required).

This is an O(N) operation, and could be fairly IO intensive.

---

## Proposed fix: Delay deleting initial_metadata_path

Currently the existence of the `initial_metadata_path` is used to denote:
_Did the job get submitted to the scheduler correctly?_

Instead its meaning could be updated to:
_Is the job in a terminal state?_

This way `Job.transition_inactive` would handle both the use cases mentioned above.

---

## Additional Fix: Migration Config File

The above fix is not currently possible as changing the meaning of `initial_metadata_path` is not backwards compatible. Any previously submitted jobs in a `pending`/`running` state would have already deleted the file.

A possible way around this is to add a new "config_file":
`~/.local/flight/job/.migration.yaml`

This would save the last known version of `flight-job` used. The next time `flight-job` is ran, it will attempt to do the following:

* Check if the config exists and is at least the current version,
* Do nothing if it is at-least the current version,
* Otherwise, loop through all the `jobs` to added the `initial_metadata_path` back (if required), and
* Update and save the migration file.

*NOTE*: Ideally the logic for the migration should not depend on the `models`. Instead the migration should be attempted syntactically. This is to prevent the migration becoming coupled to a particularly version of `Job`. This may or may not be possible.

[the remark this answers is not in the document]
Comment (benarmston) [reply lib/flight_job/commands/run_monitor.rb:33]: What I'm taking from that is that `Job.load_all` calls `Job#monitor` as a side effect. I suspect that we'll likely want to decouple that at some point.  But it should be good enough for now.

[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/flight_job/commands/run_monitor.rb:33]: Ah okay

I could add a `monitor_all` method if you like? But currently `load_all` and `monitor_all` is the same thing. It isn't possible to monitor a job without loading the metadata

**Comment by WilliamMcCumstie** [comment]
Rebased on master to resolve conflicts
